R/pairwise-comparisons.R

```diff
@@ -0,0 +1,244 @@
+  all_metrics <- list_of_avail_metrics()
+  by <- setdiff(names(scores), c(all_metrics, var))
+
+  # for efficiency, remove everything that is not in c(by, var)
+  scores <- data.table::as.data.table(scores)
+  cols_to_remove <- setdiff(names(scores), c(by, var, metric))
+  scores[, eval(cols_to_remove) := NULL]
+
+
+  single_comparison <- function(model1, model2, permutation_test = FALSE) {
+    # select only columns in c(by, var)
+    a <- scores[model == model1, ]
+    b <- scores[model == model2, ]
+
+    overlap <- merge(a, b, by = by)
+
+    #-------------------------------------------------
+    # need to do error handling if there is no overlap.
+    #-------------------------------------------------
+
+    values_x <- overlap[[paste0(metric, ".x")]]
+    values_y <- overlap[[paste0(metric, ".y")]]
+
+    # ratio should be equivalent to theta_ij in Johannes document
+    ratio <- sum(values_x) / sum(values_y)
+
+    if (permutation_test) {
+      # taken from the surveillance package
+      permutation_test <- function(score1, score2, nPermutation = 999, ratio = FALSE) {
+        nTime = length(score1)
+        meanScore1 <- mean(score1)
+        meanScore2 <- mean(score2)
+
+        if (ratio) {
+          testStat_observed <- meanScore1 / meanScore2
+        } else {
+          testStat_observed <- meanScore1 - meanScore2
+        }
+
+        testStat_permuted <- replicate(nPermutation, {
+          sel <- rbinom(nTime, size = 1, prob = 0.5)
+          g1 <- (sum(score1[sel == 0]) + sum(score2[sel == 1]))/nTime
+          g2 <- (sum(score1[sel == 1]) + sum(score2[sel == 0]))/nTime
+
+          if (ratio) {
+            g1 / g2
+          } else {
+            g1 - g2
+          }
+        })
+
+        pVal <- (1 + sum(abs(testStat_permuted) >= abs(testStat_observed)))/(nPermutation + 1)
+        # plus ones to make sure p-val is never 0?
+        return(pVal)
+      }
+
+      pval <- permutation_test(values_x, values_y)
+    } else {
+      pval <- NA
+    }
+    return(list(ratio = ratio, pval = pval))
+  }
+
+  # we only need to do the calculation once, because for the ratio that
+  # should just be the inverse and for the permuation the result should
+  # be the same
+
+  combinations <-   data.table::as.data.table(t(combn(identifiers, m = 2)))
+  colnames(combinations) <- c("model", "compare_against")
+
+  combinations[, c("ratio", "pval") := single_comparison(model, compare_against,
+                                                         permutation_test),
+               by = seq_len(NROW(combinations))]
+  combinations[sort(ratio)]
+
+  # mirror computations
+  combinations2 <- data.table::copy(combinations)
+  data.table::setnames(combinations2, old = c("model", "compare_against"),
+                       new = c("compare_against", "model"))
+  combinations2[, ratio := 1/ratio]
+
+  # add a one for those that are the same
+  combinations3 <- data.table::data.table(model = identifiers,
```

Comment:
Been reflecting slightly on the package::function usage in packages. Coming around to the idea of not using it (annoying I know) as then your functions will break if you have forgot an import whereas using this approach they don't (which is why its nice in code outside of packages). Advantage is that it is very clear where everything is coming from. Interesting to hear your thoughts.

Reply:
my thought is that running check will make sure that everything is also an import, so that would not be much of an issue. I actually like it quite a bit because it makes very clear where every function is coming from. When trying to read other people's code I like to see whether something is their own function or whether it comes from somewhere else

Reply:
Yes those were both my motivations as well. As we have discussed been struggling to get contributions to code and I was wondering if this style was off putting.

Reply:
hm.. I wouldn't say so